File: basic/agent_api.js
/**
 * RoraOS Agent Chat API - JavaScript Example
 * ===========================================
 *
 * Example of using the Agent Chat API.
 * Each agent has a unique API key and pre-configured system prompt.
 *
 * Usage in Node.js:
 *   node agent_api.js
 *
 * Usage in Browser:
 *   Import the required functions
 */

// Configuration
const API_URL = "https://labs.roraos.com/api/v1/agents/chat";
const AGENT_API_KEY = "agent_xxxxxxxxxxxxx"; // Agent-specific API key

/**
 * Send chat request to Agent API
 *
 * @param {Array} messages - Array of messages [{role, content}]
 * @param {Object} options - Additional options
 * @param {boolean} options.stream - Use streaming
 * @param {number} options.temperature - Override temperature (optional)
 * @param {number} options.max_tokens - Override max_tokens (optional)
 * @returns {Promise} Response from API
 */
async function agentChat(messages, options = {}) {
  const { stream = false, temperature, max_tokens } = options;

  const payload = {
    messages,
    stream,
  };

  // Optional overrides
  if (temperature !== undefined) payload.temperature = temperature;
  if (max_tokens !== undefined) payload.max_tokens = max_tokens;

  if (stream) {
    return agentChatStream(payload);
  }

  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${AGENT_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "API request failed");
  }

  return response.json();
}

/**
 * Handle streaming response
 */
async function agentChatStream(payload) {
  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${AGENT_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "API request failed");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let fullContent = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    const chunk = decoder.decode(value);
    const lines = chunk.split("\n");

    for (const line of lines) {
      if (line.startsWith("data: ")) {
        const data = line.slice(6);
        if (data === "[DONE]") break;

        try {
          const parsed = JSON.parse(data);
          const content = parsed.content || "";
          if (content) {
            fullContent += content;
            process.stdout.write(content);
          }
        } catch {
          // Skip invalid JSON
        }
      }
    }
  }

  console.log(); // New line
  return fullContent;
}

// ============ Usage Examples ============

async function main() {
  // Example 1: Chat with agent
  console.log("=".repeat(50));
  console.log("Example 1: Chat with Agent");
  console.log("=".repeat(50));

  try {
    const messages = [
      { role: "user", content: "Hello! What can you help me with?" },
    ];
    const result = await agentChat(messages);
    console.log(`Agent: ${result.agent?.name || "Unknown"}`);
    console.log(`Response: ${result.choices[0].message.content}`);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }

  // Example 2: Multi-turn conversation
  console.log("\n" + "=".repeat(50));
  console.log("Example 2: Multi-turn Conversation");
  console.log("=".repeat(50));

  try {
    const conversation = [];

    // Turn 1
    conversation.push({
      role: "user",
      content: "I want to learn programming",
    });

    let result = await agentChat(conversation);
    const response1 = result.choices[0].message.content;
    console.log(`User: ${conversation[conversation.length - 1].content}`);
    console.log(`Agent: ${response1}`);

    // Save response
    conversation.push({ role: "assistant", content: response1 });

    // Turn 2
    conversation.push({ role: "user", content: "Where should I start?" });

    result = await agentChat(conversation);
    console.log(`\nUser: ${conversation[conversation.length - 1].content}`);
    console.log(`Agent: ${result.choices[0].message.content}`);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }

  // Example 3: Streaming with override settings
  console.log("\n" + "=".repeat(50));
  console.log("Example 3: Streaming with Custom Settings");
  console.log("=".repeat(50));

  try {
    const messages = [
      { role: "user", content: "Give me productivity tips." },
    ];
    process.stdout.write("Response: ");
    await agentChat(messages, {
      stream: true,
      temperature: 0.8,
      max_tokens: 500,
    });
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }
}

// Run examples
main().catch(console.error);

// Export for use as module
if (typeof module !== "undefined") {
  module.exports = { agentChat, agentChatStream };
}

File: basic/agent_api.py
"""
RoraOS Agent Chat API - Python Example
=======================================

Example of using the Agent Chat API.
Each agent has a unique API key and pre-configured system prompt.

Installation:
    pip install requests

Usage:
    python agent_api.py
"""

import requests
import json

# Configuration
API_URL = "https://labs.roraos.com/api/v1/agents/chat"
AGENT_API_KEY = "agent_xxxxxxxxxxxxx"  # Agent-specific API key (get from agent page)


def agent_chat(messages: list, stream: bool = False, temperature: float = None, max_tokens: int = None):
    """
    Send chat request to Agent API.

    The agent already has a system prompt, knowledge base, and its own configuration.

    Args:
        messages: List of messages with format [{"role": "user", "content": "Hello"}]
        stream: Use streaming response
        temperature: Override agent temperature (optional)
        max_tokens: Override agent max_tokens (optional)

    Returns:
        Response from API
    """
    headers = {
        "Authorization": f"Bearer {AGENT_API_KEY}",
        "Content-Type": "application/json"
    }

    payload = {
        "messages": messages,
        "stream": stream
    }

    # Optional overrides
    if temperature is not None:
        payload["temperature"] = temperature
    if max_tokens is not None:
        payload["max_tokens"] = max_tokens

    if stream:
        return agent_chat_stream(headers, payload)

    response = requests.post(API_URL, headers=headers, json=payload)
    response.raise_for_status()
    return response.json()


def agent_chat_stream(headers: dict, payload: dict):
    """Handle streaming response."""
    response = requests.post(
        API_URL,
        headers=headers,
        json=payload,
        stream=True
    )
    response.raise_for_status()

    full_content = ""
    for line in response.iter_lines():
        if line:
            line = line.decode('utf-8')
            if line.startswith('data: '):
                data = line[6:]
                if data == '[DONE]':
                    break
                try:
                    chunk = json.loads(data)
                    content = chunk.get('content', '')
                    if content:
                        full_content += content
                        print(content, end='', flush=True)
                except json.JSONDecodeError:
                    pass

    print()  # New line after streaming
    return full_content


# ============ Usage Examples ============

if __name__ == "__main__":
    # Example 1: Chat with agent
    print("=" * 50)
    print("Example 1: Chat with Agent")
    print("=" * 50)

    messages = [
        {"role": "user", "content": "Hello! What can you help me with?"}
    ]

    try:
        result = agent_chat(messages)
        print(f"Agent: {result['agent']['name']}")
        print(f"Response: {result['choices'][0]['message']['content']}")
    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")

    # Example 2: Multi-turn conversation
    print("\n" + "=" * 50)
    print("Example 2: Multi-turn Conversation")
    print("=" * 50)

    conversation = []

    # Turn 1
    conversation.append({"role": "user", "content": "I want to learn programming"})

    try:
        result = agent_chat(conversation)
        assistant_response = result['choices'][0]['message']['content']
        print(f"User: {conversation[-1]['content']}")
        print(f"Agent: {assistant_response}")

        # Save response for context
        conversation.append({"role": "assistant", "content": assistant_response})

        # Turn 2
        conversation.append({"role": "user", "content": "Where should I start?"})

        result = agent_chat(conversation)
        print(f"\nUser: {conversation[-1]['content']}")
        print(f"Agent: {result['choices'][0]['message']['content']}")

    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")

    # Example 3: Streaming with override settings
    print("\n" + "=" * 50)
    print("Example 3: Streaming with Custom Settings")
    print("=" * 50)

    messages = [
        {"role": "user", "content": "Give me productivity tips."}
    ]

    try:
        print("Response: ", end='')
        agent_chat(messages, stream=True, temperature=0.8, max_tokens=500)
    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")
